Add support for AssumeRoleWithWebIdentity

devkid opened the pull request:
*Description of changes:*
This allows to provide credentials for mounting IAM-restricted EFS filesystems by using IRSA (IAM Role via ServiceAccount) on EKS.

Closes #54.

wochanda commented:
Thanks for the request! Could you say more about how the role would be used in EKS, and how you expect it to interact with the EFS CSI driver? 

AFAIK only the user pod itself has access to the WebIdentity, not the CSI driver pod, so this would require you to make the mount from the container itself, rather than using the CSI driver. Is that right?

devkid commented:
The AWS EFS CSI driver is running as a DaemonSet and is performing the mount. The idea is that you can assign an IAM role to the CSI driver pods and only allow this role to access a specific EFS file system (via filesystem policy).

I have tested this in our cluster and it works as expected.

devkid commented:
I just noticed that the watchdog also needs to have the same implementation. It worked for my initial testings with mounting new volumes etc. but fails if the host already had a mounted volume and the daemonset was restarted. I'm on it.

devkid commented:
Changes for the watchdog are also done now.

Cappuccinuo commented on src/watchdog/__init__.py:
```diff
@@ -185,6 +191,42 @@ def get_aws_security_credentials_from_ecs(uri):
+        with open(token_file, 'r') as f:
+            token = f.read()
+    except Exception as e:
+        if is_fatal:
```
`is_fatal` is not defined

Cappuccinuo replied:
Just pass `is_fatal=False` to the func is OK.

devkid replied:
I removed `is_fatal` altogether and now just print the error (as the other methods in the watchdog also don't have it).

Cappuccinuo commented on src/watchdog/__init__.py:
```diff
@@ -185,6 +191,42 @@ def get_aws_security_credentials_from_ecs(uri):
+    resp = url_request_helper(webidentity_url, unsuccessful_resp, url_error_msg, headers={'Accept': 'application/json'})
+
+    if resp:
+        creds = resp.get('AssumeRoleWithWebIdentityResponse', {}).get('AssumeRoleWithWebIdentityResult', {}).get('Credentials', {})
```
This line is too long.

devkid replied:
Fixed. I didn't find any style guidelines. Feel free to change it again if you don't like the new format. (Also there are longer lines in the existing code, so this isn't an indication either.)

Cappuccinuo commented on src/mount_efs/__init__.py:
```diff
@@ -329,6 +335,11 @@ def get_aws_security_credentials(use_iam, awsprofile=None, aws_creds_uri=None):
             return credentials, credentials_source
 
+    if WEB_IDENTITY_ROLE_ARN_ENV in os.environ and WEB_IDENTITY_TOKEN_FILE_ENV in os.environ:
+        credentials, credentials_source = get_aws_security_credentials_from_webidentity(os.environ[WEB_IDENTITY_ROLE_ARN_ENV], os.environ[WEB_IDENTITY_TOKEN_FILE_ENV], False)
```
This line is too long.

Cappuccinuo commented on src/mount_efs/__init__.py:
```diff
@@ -376,6 +387,46 @@ def get_aws_security_credentials_from_ecs(aws_creds_uri, is_fatal=False):
+    resp = url_request_helper(webidentity_url, unsuccessful_resp, url_error_msg, headers={'Accept': 'application/json'})
+
+    if resp:
+        creds = resp.get('AssumeRoleWithWebIdentityResponse', {}).get('AssumeRoleWithWebIdentityResult', {}).get('Credentials', {})
```
This line is too long.

devkid commented:
@kbasv Ready for review. We have this running in our clusters for a few weeks already and it's working for us.

kbasv commented on src/mount_efs/__init__.py:
```diff
@@ -329,6 +335,15 @@ def get_aws_security_credentials(use_iam, awsprofile=None, aws_creds_uri=None):
         if credentials and credentials_source:
             return credentials, credentials_source
 
+    if WEB_IDENTITY_ROLE_ARN_ENV in os.environ and WEB_IDENTITY_TOKEN_FILE_ENV in os.environ:
```
Can you add a comment above on how this piece of code attempts to retrieve credentials?

devkid replied:
Done.

devkid commented:
@kbasv Tests are fixed.

Cappuccinuo commented:
@devkid Overall this looks good to me. Could you please rebase all of your commits into a single commit before we merge please?

stvhu commented:
@devkid, just want to echo @Cappuccinuo, if you may squash down your commits, but other than that, looks good on my end!

devkid commented:
@Cappuccinuo @stvhu done.

Cappuccinuo commented:
Thanks @devkid.

The fedora failure can be ignored, the reason is the new pulled docker image does not have systemd installed. LGTM.

devkid commented:
Is there anything left that I need to do go this merged?

Cappuccinuo commented:
@devkid We'll do some additional testing on our side before merging the pull request, will let you know if anything else is needed. Thanks.

kbasv commented:
@devkid Can you provide details on how you were able to test this change on your cluster?

devkid commented:
@kbasv I created an IAM role with no permissions and added a trust policy to allow it to be assumed by the service account of the AWS EFS CSI driver daemonset via IRSA.

I added a service account of the daemonset and added the `eks.amazonaws.com/role-arn` annotation:

```yaml
apiVersion: v1
kind: ServiceAccount
metadata:
  name: efs-csi-node
  annotations:
    eks.amazonaws.com/role-arn: arn:aws:iam::1234567890:role/aws-efs-csi-driver
---
kind: DaemonSet
apiVersion: apps/v1
namespace:
  name: efs-csi-node
# ...
spec:
  # ...
  template:
    # ...
    spec:
      serviceAccountName: efs-csi-node
      # ...
```

In EFS, I added the following policy to the file system:
```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Sid": "efs-statement-19e6d5c0-69f5-49f9-afe2-31793e92f8e9",
            "Effect": "Allow",
            "Principal": {
                "AWS": [
                    "arn:aws:iam::1234567890:role/aws-efs-csi-driver"
                ]
            },
            "Action": [
                "elasticfilesystem:ClientMount",
                "elasticfilesystem:ClientWrite",
                "elasticfilesystem:ClientRootAccess"
            ],
            "Resource": "arn:aws:elasticfilesystem:eu-central-1:1234567890:file-system/fs-fa3498a2"
        }
    ]
}
```

esalberg commented:
After rebuilding amazon-efs-utils with this patch, we confirmed that AssumeRoleWithIdentity for efs-mount-helper does assume the app role per CloudTrail (see ID 1acf7d09-2493-409c-9d5a-c742caa4d282 ).  We can mount the protected EFS share manually in a pod using the “mount” command (did not work before the amazon-efs-utils patch).

However, a pod is not able to deploy with a volumeMount specifying a protected share.  CloudTrail (see ID 93caf173-8c0c-4fe4-973d-9c40469c5623 ) shows that NewClientConnection is trying to use the EKS worker node role, not the app role:

    "userIdentity": {
        "type": "AssumedRole",
        "principalId": "<snip>:i-069489e612bf1c8f9",
        "arn": "arn:aws:sts::932338283021:assumed-role/eksctl-deveks01-nodegroup-worker-NodeInstanceRole-O7F9X42KE4HX/i-069489e612bf1c8f9",
<snip>
 "eventSource": "elasticfilesystem.amazonaws.com",
    "eventName": "NewClientConnection",
    "awsRegion": "us-east-1",
    "sourceIPAddress": "AWS Internal",
    "userAgent": "elasticfilesystem",
    "errorCode": "AccessDenied",

Did you see anything like this in your testing?

Mount command that works inside the pod after patching amazon-efs-utils with this PR:
mount -t efs -o tls,iam,accesspoint=fsap-<fasp_id> fs-<fs_id> /mnt

PV info for the PVC that doesn't mount as part of the deployment:
    VolumeHandle:      fs-<fs_id>:/persistentvolumes/mymount-pvc-<id>:fsap-<fasp_id>

Snippets from the deployment yaml, just to be complete:
    volumeMounts:
    - mountPath: /efs
      name: mymount

      volumes:
      - name: mymount
        persistentVolumeClaim:
          claimName: mymount

kbasv commented:
@esalberg By `Protected EFS share` do you mean access points?

I tested access-points mount (both manual & via deployment) and CSI driver is using the app role and mounts are successful. Cloud Trail events indicate efs-mount-helper is using the app role associated with the service account for `NewClientConnection`
 
Can you provide the following details?
1. Service account setup
2. IAM policy associated with the role
3. mount.log from when the deployment failed.

Thanks!

esalberg commented:
> @esalberg By `Protected EFS share` do you mean access points?

Yes, we are using access points with filesystem policy to limit access to a specific role.
 
> I tested access-points mount (both manual & via deployment) and CSI driver is using the app role and mounts are successful. Cloud Trail events indicate efs-mount-helper is using the app role associated with the service account for `NewClientConnection`
> 
> Can you provide the following details?
> 
> 1. Service account setup
> 2. IAM policy associated with the role
> 3. mount.log from when the deployment failed.

That is very interesting.  We are working with AWS support, and they stated that there is a limitation in the EFS CSI driver:

"The issue seems to stem from the EFS CSI driver not passing the proper IAM flag to the underlying efs-utils package that the driver utilizes. This has been reported on the CSI driver repository, as well as the AWS container roadmap as seen:
https://github.com/kubernetes-sigs/aws-efs-csi-driver/issues/124
https://github.com/aws/containers-roadmap/issues/1003

 EFS Utils requires passing the flag -o iam to the mount command to properly pass IAM roles for authorization. As the EFS CSI driver is solely responsible for fulling Kubernetes Persistent Volume requests, it does not currently read Kubernetes Service Account information from pods to facilitate authorization.

You can see documentation for the IAM flag in the announcement blog post for EFS Access Points here. An example working mount command would be as follows:
$ sudo mount -t efs -o iam,tls,accesspoint=fsap-0204ce67a2208742e fs-d1188b58 /mnt/shared

From my testing, looking at the logs of the EFS CSI node driver, you can see the mount command being passed here, without the required IAM flag:
2020-08-07T22:05:30.081026058Z Mounting arguments: -t efs -o accesspoint=fsap-0e706c8f29242f0d1,tls fs-7418ecf6:/ /var/lib/kubelet.pods/dde3c956-2d4e-4e80-b656-dcb282e6289c/volumes/kubernetes.io~csi/efs-pv-2/mount

2020-08-07T22:05:30.08103621Z Output: Could not start amazon-efs-mount-watchdog, unrecognized init system “aws-efs-csi-dri”

2020-08-07T22:05:30.081033998Z mount.nfs4: access denied by server while mounting 127.0.0.1:/

Since the flag is not passed, the driver does not try to read any IAM role, and is therefore unable to assume the role required to mount the access point with policy attached. The relevant code that is called inside the driver is located here, where the flag is not passed: https://github.com/kubernetes-sigs/aws-efs-csi-driver/blob/b9d26737c3f8659403eaf753a61c7874ca4a8b39/pkg/driver/node.go#L114

esalberg commented:
> Service account setup:

apiVersion: v1
kind: ServiceAccount
metadata:
  annotations:
    eks.amazonaws.com/role-arn: arn:aws:iam::932338283021:role/apps.jenkins
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"v1","kind":"ServiceAccount","metadata":{"annotations":{"eks.amazonaws.com/role-arn":"arn:aws:iam::932338283021:role/apps.jenkins"},"name":"default","namespace":"jenkins"}}
  creationTimestamp: "2020-07-20T13:13:00Z"
  name: default
  namespace: jenkins
  resourceVersion: "5565332"
  selfLink: /api/v1/namespaces/jenkins/serviceaccounts/default
  uid: <uid_snip>
secrets:
- name: default-token-ft75x


> IAM policy associated with the role:
This is not easy to export / share.  I will note that this policy works file for an EFS access point with filesystem policy from EC2, as well as secrets access with IRSA.

Trusted entities:

The identity provider(s) events.amazonaws.com
--
The identity provider(s) states.amazonaws.com
The identity provider(s) eks.amazonaws.com
The identity provider(s) lambda.amazonaws.com
The identity provider(s) elasticfilesystem.amazonaws.com
The identity provider(s) ec2.amazonaws.com
arn:aws:iam::932338283021:oidc-provider/oidc.eks.us-east-1.amazonaws.com/id/<oidc_id>

Conditions:
StringEquals	oidc.eks.us-east-1.amazonaws.com/id/<oidc_id>:aud	
sts.amazonaws.com

StringLike	oidc.eks.us-east-1.amazonaws.com/id/<oidc_id>:sub	
system:serviceaccount:jenkins:*

system:serviceaccount:jenkins*:*

        {
            "Sid": "apps-jenkins-ap",
            "Effect": "Allow",
            "Principal": {
                "AWS": [
                    "arn:aws:iam::932338283021:role/apps.jenkins"
                ]
            },
            "Action": [
                "elasticfilesystem:ClientMount",
                "elasticfilesystem:ClientWrite",
                "elasticfilesystem:ClientRootAccess"
            ],
            "Resource": "arn:aws:elasticfilesystem:us-east-1:932338283021:file-system/fs-<id>",
            "Condition": {
                "StringEquals": {
                    "elasticfilesystem:AccessPointArn": "arn:aws:elasticfilesystem:us-east-1:932338283021:access-point/fsap-<ap_id>"
                }
            }
        }

(We also tried adding just in case: "arn:aws:sts::932338283021:assumed-role/apps.jenkins/efs-mount-helper", - did not help)

> mount.log from when the deployment failed:
The deployment hung and wouldn't get to the point where we could get the mount.log.  However, you can see above for what AWS support reported in their log.

kbasv commented:
> **Were you passing the role to the EFS CSI driver specifically, or were you passing the role from the pod itself?**

I was passing the role to the CSI driver in my testing as CSI driver is responsible for performing the mounts (all mounts need to have access to this role )

esalberg commented:
Thanks for the confirmation - two different use cases, then.  We are able to pass a different role than the worker node role for all mounts (efs csi pod), and we will follow up with AWS for passing multiple Mount roles for multiple pods.

Sent from my iPhone

On Aug 14, 2020, at 6:14 PM, Karthik Basavaraj <notifications@github.com> wrote:

﻿

Were you passing the role to the EFS CSI driver specifically, or were you passing the role from the pod itself?

I was passing the role to the CSI driver in my testing as CSI driver is responsible for performing the mounts (all mounts need to have access to this role )

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_aws_efs-2Dutils_pull_60-23issuecomment-2D674293503&d=DwMCaQ&c=vYgbkq6q_4hmkgETKofs3A&r=dYP_6b0bY8kFzOVJ7ZLYv2UB6hf7GylA08dLczMmM-w&m=rL74urmOD7TuJbGC582Jq3N1ebc8Z4y3D6DDCF6eYZE&s=2-L-xeuSba69PZErIjnM-kimOa2M9pzeYlTa9PXCWck&e=>, or unsubscribe<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_ACHPV4VI6LZS6QZ6DAFEJ73SAWZNLANCNFSM4NBSMIMQ&d=DwMCaQ&c=vYgbkq6q_4hmkgETKofs3A&r=dYP_6b0bY8kFzOVJ7ZLYv2UB6hf7GylA08dLczMmM-w&m=rL74urmOD7TuJbGC582Jq3N1ebc8Z4y3D6DDCF6eYZE&s=8SaDZKIfbcCa6OO7UenF9_1jRQVMzBrLIG-KaugrXSM&e=>.


This message is intended for the recipient only and is not meant to be forwarded or distributed in any other format. This communication is for informational purposes only. It is not intended as an offer or solicitation for the purchase or sale of any financial instrument, or security, or as an official confirmation of any transaction. Putnam does not accept purchase or redemptions of securities, instructions, or authorizations that are sent via e-mail. All market prices, data and other information are not warranted as to completeness or accuracy and are subject to change without notice. Any comments or statements made herein do not necessarily reflect those of Putnam Investments, LLC (DBA Putnam Investments) and its subsidiaries and affiliates. If you are not the intended recipient of this e-mail, please delete the e-mail. Information on how Putnam gathers and uses personal data is available in Putnam’s privacy notice which can be found at https://www.putnam.com/policies

daisuke-yoshimoto commented:
@kbasv @Cappuccinuo @stvhu 
In addition to the content of this PR, I fixed to use regional AWS STS endpoints instead of the global endpoint to reduce latency.  Can you please review this PR? https://github.com/aws/efs-utils/pull/84
